sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,277 @@
+    return self._position
+
+  def seekable(self):
+    return True
+
+  def readable(self):
+    return True
+
+
+class UploaderStream(io.RawIOBase):
+  """Provides a stream interface for Downloader objects."""
+
+  def __init__(self, uploader, mode='w'):
+    """Initializes the stream.
+
+    Args:
+      uploader: (Uploader) Filesystem dependent implementation.
+      mode: (string) Python mode attribute for this stream.
+    """
+    self._uploader = uploader
+    self.mode = mode
+    self._position = 0
+
+  def write(self, b):
+    """Write bytes from b.
+
+    Returns number of bytes written (<= len(b)).
+
+    Args:
+      b: (memoryview) Buffer with data to write.
+    """
+    self._checkClosed()
+    try:
+      self._uploader.put(b)
+    except IOError:
+      # TODO(udim): Call self._uploader.finish() instead (once
+      # GcsUploader.finish() calls join() with timeout), and remove last_error
+      # property. Rely on finish() to raise any internal errors encountered.
+      if self._uploader.last_error:
+        raise self._uploader.last_error  # pylint: disable=raising-bad-type
+      else:
+        raise
+
+    bytes_written = len(b)
+    self._position += bytes_written
+    return bytes_written
+
+  def close(self):
+    """Complete the upload and close this stream.
+
+    This method has no effect if the stream is already closed.
+
+    Raises:
+      Any error encountered by the uploader.
+    """
+    if not self.closed:
+      self._uploader.finish()
+
+    super(UploaderStream, self).close()
+
+  def writable(self):
+    return True
+
+
+class PipeStream(object):
+  """A class that presents a pipe connection as a readable stream."""
+
+  def __init__(self, recv_pipe):
+    self.conn = recv_pipe
+    self.closed = False
+    self.position = 0
+    self.remaining = ''
+
+  def read(self, size):
+    """Read data from the wrapped pipe connection.
+
+    Args:
+      size: Number of bytes to read. Actual number of bytes read is always
+            equal to size unless EOF is reached.
+
+    Returns:
+      data read as str.
+    """
+    data_list = []
+    bytes_read = 0
+    while bytes_read < size:
+      bytes_from_remaining = min(size - bytes_read, len(self.remaining))
+      data_list.append(self.remaining[0:bytes_from_remaining])
+      self.remaining = self.remaining[bytes_from_remaining:]
+      self.position += bytes_from_remaining
+      bytes_read += bytes_from_remaining
+      if not self.remaining:
+        try:
+          self.remaining = self.conn.recv_bytes()
+        except EOFError:
+          break
+    return ''.join(data_list)
+
+  def tell(self):
+    """Tell the file's current offset.
+
+    Returns:
+      current offset in reading this file.
+
+    Raises:
+      ``ValueError``: When this stream is closed.
+    """
+    self._check_open()
+    return self.position
+
+  def seek(self, offset, whence=os.SEEK_SET):
+    # The gcsio.Uploader class insists on seeking to the end of a stream to
```

Comment:
<!--new_thread; commit:79815e6df1d49c271a6c44e37fb2a3c46be0aed5; resolved:0-->
"The apitools library used by the gcsio.Uploader class"